### PR TITLE
[gh-791] move coin_id to multichain_id.

### LINKED ISSUE #791: [multichain-ids] build custom chain ids to separate from built-in chain ids
Discussion:

https://github.com/rooch-network/rooch/pull/787#discussion_r1322428906
https://discord.com/channels/1078938449974935592/1151268013169979486

Build custom chain ids for each blockchain to substitute `coin_id` and separate from built-in Rooch chain ids.

I think we can remove `CoinID` from `coin_type.rs` and build from CustomChainID here:

https://github.com/rooch-network/rooch/blob/9ebd7c3b1ba61fcc83ce278ae92118f2fb97e580/crates/rooch-types/src/chain_id.rs#L197-L210

### PR DESCRIPTION
**Pull request by feliciss**
## Summary

Following discussions on PR #787.

1. Delete CoinID.
2. Create `multichain_id` to use with RoochMultiChainID with built-in ids.

- Closes #791

### CODE REVIEW

**Comment by jolestar** [comment]
the custom chain id is for launching a new rooch network, the usage scenarios is different with coin id.

**Comment by feliciss** [comment]
> the custom chain id is for launching a new rooch network, the usage scenarios is different with coin id.

Okay. Should there be an another file named `multichain_id.rs`?

File: crates/rooch-key/src/key_derive.rs [hunk 1]
```diff
@@ -11,13 +11,13 @@ use fastcrypto::secp256k1::recoverable::{
...
 use rooch_types::error::RoochError;
+use rooch_types::multichain_id::RoochMultiChainID;
 use slip10_ed25519::derive_ed25519_private_key;
 use std::string::String;
 
-// CoinID type
+// Coin type
 pub const DERIVATION_PATH_COIN_TYPE_BTC: u64 = 0;
 pub const DERIVATION_PATH_COIN_TYPE_ETH: u64 = 60;
 pub const DERIVATION_PATH_COIN_TYPE_SUI: u64 = 784;
```
Comment (baichuan3) [inline crates/rooch-key/src/key_derive.rs:23]: Here is also MultiChain ID?

File: crates/rooch-rpc-api/src/jsonrpc_types/execute_tx_response.rs [hunk 2]
```diff
@@ -80,14 +80,14 @@ impl From<KeptVMStatus> for KeptVMStatusView {
 
 #[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
 pub struct AuthenticatorView {
-    pub coin_id: StrView<u64>,
+    pub scheme: StrView<u64>,
```
Comment (baichuan3) [inline crates/rooch-rpc-api/src/jsonrpc_types/execute_tx_response.rs:83]: Typo `scheme` -> `schema`

**Review by baichuan3 — APPROVED** [review]

File: crates/rooch-rpc-api/src/jsonrpc_types/execute_tx_response.rs (hunk 2 above)
Comment (feliciss) [reply crates/rooch-rpc-api/src/jsonrpc_types/execute_tx_response.rs:83]: > Typo `scheme` -> `schema`

I believe it's `scheme` before.

File: crates/rooch-key/src/key_derive.rs (hunk 1 above)
Comment (feliciss) [reply crates/rooch-key/src/key_derive.rs:23]: > Here is also MultiChain ID?

Yep.

File: crates/rooch-framework/sources/transaction_validator.move [hunk 3]
```diff
@@ -44,6 +44,8 @@ module rooch_framework::transaction_validator {
         scheme: u64,
         authenticator_payload: vector<u8>
     ): TxValidateResult {
+        std::debug::print(&chain_id);
```
Comment (jolestar) [inline crates/rooch-framework/sources/transaction_validator.move:47]: remove debug

File: crates/rooch-rpc-api/src/jsonrpc_types/execute_tx_response.rs (hunk 2 above)
Comment (jolestar) [reply crates/rooch-rpc-api/src/jsonrpc_types/execute_tx_response.rs:83]: scheme

File: crates/rooch/src/commands/account/commands/update.rs [hunk 4]
```diff
@@ -25,67 +25,59 @@ pub struct UpdateCommand {
     mnemonic_phrase: String,
     #[clap(flatten)]
     pub context_options: WalletContextOptions,
-    /// Command line input of coin ids
-    #[clap(short = 'c', long = "coin-id", arg_enum)]
-    pub coin_id: CoinID,
+    /// Command line input of multichain ids
+    #[clap(short = 'i', long = "multichain-id")]
+    pub multichain_id: RoochMultiChainID,
```
Comment (jolestar) [inline crates/rooch/src/commands/account/commands/update.rs:30]: We do not need to support multichain_id for update command. Just remove it.

File: crates/rooch/src/commands/account/commands/nullify.rs [hunk 5]
```diff
@@ -26,59 +26,55 @@ pub struct NullifyCommand {
     address: String,
     #[clap(flatten)]
     pub context_options: WalletContextOptions,
-    /// Command line input of coin ids
-    #[clap(short = 'c', long = "coin-id", arg_enum)]
-    pub coin_id: CoinID,
+    /// Command line input of multichain ids
+    #[clap(short = 'i', long = "multichain-id")]
+    pub multichain_id: RoochMultiChainID,
```
Comment (jolestar) [inline crates/rooch/src/commands/account/commands/nullify.rs:31]: I think we can remove the multichain id here. It should be the validator sheme id, but not every validator scheme have auth key. So, just remove and keep simple.

File: crates/rooch-types/src/framework/native_validator.rs [hunk 6]
```diff
@@ -16,8 +16,8 @@ use moveos_types::{
 pub struct NativeValidator {}
 
 impl NativeValidator {
-    pub fn coin_id() -> CoinID {
-        CoinID::Rooch
+    pub fn multichain_id() -> RoochMultiChainID {
```
Comment (jolestar) [inline crates/rooch-types/src/framework/native_validator.rs:19]: This shoud be the auth validator id

File: crates/rooch-types/src/framework/ethereum_validator.rs [hunk 7]
```diff
@@ -16,8 +16,8 @@ use moveos_types::{
 pub struct EthereumValidator {}
 
 impl EthereumValidator {
-    pub fn coin_id() -> CoinID {
-        CoinID::Ether
+    pub fn multichain_id() -> RoochMultiChainID {
```
Comment (jolestar) [inline crates/rooch-types/src/framework/ethereum_validator.rs:19]: This is the auth validator id

File: crates/rooch-types/src/transaction/authenticator.rs [hunk 8]
```diff
@@ -34,8 +34,8 @@ pub struct RoochAuthenticator {
 }
 
 impl BuiltinAuthenticator for RoochAuthenticator {
-    fn coin_id(&self) -> CoinID {
-        CoinID::Rooch
+    fn scheme(&self) -> u64 {
+        BuiltinScheme::Ed25519.flag().into()
     }
```
Comment (jolestar) [inline crates/rooch-types/src/transaction/authenticator.rs:39]: We can rename the builtin scheme to builtin auth validator.

File: crates/rooch-rpc-api/src/jsonrpc_types/execute_tx_response.rs (hunk 2 above)
Comment (jolestar) [reply crates/rooch-rpc-api/src/jsonrpc_types/execute_tx_response.rs:83]: Or we rename it to authentication validator id.

File: crates/rooch-types/src/transaction/ethereum.rs [hunk 9]
```diff
@@ -50,7 +48,9 @@ impl EthereumTransactionData {
             access_list: None,
             max_priority_fee_per_gas: None,
             max_fee_per_gas: None,
-            chain_id: Some(U256::from(RoochChainID::DEV.chain_id().id())),
+            chain_id: Some(U256::from(
+                RoochMultiChainID::Rooch.multichain_id().id(), // TODO use Ethereum multichain id and build a multichain genesis init
```
Comment (jolestar) [inline crates/rooch-types/src/transaction/ethereum.rs:52]: This chain id is not the multichain id

File: crates/rooch-key/src/key_derive.rs [hunk 10]
```diff
@@ -81,95 +81,95 @@ impl CoinOperations<EthereumAddress, Secp256k1RecoverableKeyPair> for CoinID {
...
         }
-        CoinID::Bitcoin => todo!(),
-        CoinID::Nostr => todo!(),
     }
 }
 
 pub fn generate_new_key_pair<Addr, KeyPair>(
-    coin_id: CoinID,
+    multichain_id: RoochMultiChainID,
     derivation_path: Option<DerivationPath>,
```
Comment (jolestar) [inline crates/rooch-key/src/key_derive.rs:155]: We do not need to support every chain's keypair. We can replace the multichain id with a key pair type.

**Review by jolestar — CHANGES_REQUESTED** [review]

File: crates/rooch-framework/sources/transaction_validator.move (hunk 3 above)
Comment (feliciss) [reply crates/rooch-framework/sources/transaction_validator.move]: Removed.

File: crates/rooch/src/commands/account/commands/update.rs (hunk 4 above)
Comment (feliciss) [reply crates/rooch/src/commands/account/commands/update.rs]: Removed.

File: crates/rooch/src/commands/account/commands/nullify.rs (hunk 5 above)
Comment (feliciss) [reply crates/rooch/src/commands/account/commands/nullify.rs]: Removed.

File: crates/rooch-types/src/framework/ethereum_validator.rs (hunk 7 above)
Comment (feliciss) [reply crates/rooch-types/src/framework/ethereum_validator.rs]: Changed to auth_validator_id().

File: crates/rooch-types/src/framework/native_validator.rs (hunk 6 above)
Comment (feliciss) [reply crates/rooch-types/src/framework/native_validator.rs]: Changed to auth_validator_id().

File: crates/rooch-types/src/transaction/authenticator.rs (hunk 8 above)
Comment (feliciss) [reply crates/rooch-types/src/transaction/authenticator.rs:39]: Renamed BuiltinAuthValidator.

File: crates/rooch-types/src/transaction/ethereum.rs (hunk 9 above)
Comment (feliciss) [reply crates/rooch-types/src/transaction/ethereum.rs]: Currently I changed to `20230103` in default genesis build. Hopefully I will build ethereum chain id on genesis init later.

File: crates/rooch-key/src/key_derive.rs (hunk 10 above)
Comment (feliciss) [reply crates/rooch-key/src/key_derive.rs:149]: I defined a `KeyPairType` and assign every key pairs a type either of `evm` or `rooch`.

File: crates/rooch-key/src/key_derive.rs (hunk 1 above)
Comment (feliciss) [reply crates/rooch-key/src/key_derive.rs]: Parsed them from `multichain_id.rs`.

File: crates/rooch-rpc-api/src/jsonrpc_types/execute_tx_response.rs (hunk 2 above)
Comment (feliciss) [reply crates/rooch-rpc-api/src/jsonrpc_types/execute_tx_response.rs]: Renamed `scheme` to `auth_validator` for BuiltinAuthValidator and `auth_validator_id` for u64 ids.

File: crates/rooch-types/proptest-regressions/transaction/authenticator.txt
Comment (jolestar) [inline crates/rooch-types/proptest-regressions/transaction/authenticator.txt:1]: Is there a proptest failure case?

File: crates/rooch-types/src/address.rs [hunk 11]
```diff
@@ -44,24 +48,26 @@ pub trait RoochSupportedAddress:
...
         Ok(Self {
-            coin_id,
+            multichain_id,
             raw_address,
         })
     }
 
     pub fn is_rooch_address(&self) -> bool {
-        self.coin_id == CoinID::Rooch
+        self.multichain_id.multichain_id().id() == BuiltinChainID::Dev.chain_id().id()
```
Comment (jolestar) [inline crates/rooch-types/src/address.rs:68]: Rooch only has one multichain_id. We do not need to distinguish dev|test|main in the multichain_id scenario.

File: crates/rooch-types/src/multichain_id.rs [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+use std::fmt::{Display, Formatter};
+use std::str::FromStr;
+
+use crate::chain_id::ChainID;
+
+pub const BITCOIN: u64 = 0;
+pub const ETHER: u64 = 60;
+pub const SUI: u64 = 784;
+pub const NOSTR: u64 = 1237;
+pub const ROOCH: u64 = 20230103; // Dev
```
Comment (jolestar) [inline crates/rooch-types/src/multichain_id.rs:19]: Use main chain_id as multichain_id. We can also use another id here, using the same value as chain_id just for easy to remember.

File: crates/rooch-types/src/transaction/ethereum.rs [hunk 13]
```diff
@@ -50,7 +48,8 @@ impl EthereumTransactionData {
             access_list: None,
             max_priority_fee_per_gas: None,
             max_fee_per_gas: None,
-            chain_id: Some(U256::from(RoochChainID::DEV.chain_id().id())),
+            // TODO create ethereum chain id in genesis init to pass through transaction validator tests
```
Comment (jolestar) [inline crates/rooch-types/src/transaction/ethereum.rs:51]: We do not need to define an ethereum chain id, we just use the RoochChainID.

File: crates/rooch-types/src/transaction/ethereum.rs [hunk 14]
```diff
@@ -155,10 +154,7 @@ impl AbstractTransaction for EthereumTransactionData {
 
     fn authenticator_info(&self) -> Result<AuthenticatorInfo> {
         let chain_id = self.0.chain_id.ok_or(RoochError::InvalidChainID)?.as_u64();
-        let authenticator = Authenticator::new(
-            CoinID::Ether as u64,
-            self.into_signature()?.as_bytes().to_vec(),
-        );
+        let authenticator = Authenticator::new(1, self.into_signature()?.as_bytes().to_vec());
```
Comment (jolestar) [inline crates/rooch-types/src/transaction/ethereum.rs:157]: We need to define an Enum for `1`. BuiltinAuthValidator::Ethereum?. We should keep the Ethereum signature in its original format.

**Review by jolestar — APPROVED** [review]
Some refactoring can be done later.

File: crates/rooch-types/src/transaction/ethereum.rs (hunk 14 above)
Comment (feliciss) [reply crates/rooch-types/src/transaction/ethereum.rs]: Defined and used Ethereum signature from `ethers` lib.

File: crates/rooch-types/src/transaction/ethereum.rs (hunk 13 above)
Comment (feliciss) [reply crates/rooch-types/src/transaction/ethereum.rs]: Resolved to RoochChainID DEV.

File: crates/rooch-types/src/multichain_id.rs (hunk 12 above)
Comment (feliciss) [reply crates/rooch-types/src/multichain_id.rs]: I gave it to a slip-0044 placeholder. Now it's 20230101.

File: crates/rooch-types/src/address.rs (hunk 11 above)
Comment (feliciss) [reply crates/rooch-types/src/address.rs]: Resolved to `is_rooch`.

File: crates/rooch-types/proptest-regressions/transaction/authenticator.txt
Comment (feliciss) [reply crates/rooch-types/proptest-regressions/transaction/authenticator.txt:1]: Sorry it was a previous failure I'd included it. Now it's gone.

File: crates/rooch-framework/sources/crypto/ed25519.move [hunk 15]
```diff
@@ -21,8 +21,8 @@ module rooch_framework::ed25519 {
 
     public fun get_public_key_from_authenticator_payload(authenticator_payload: &vector<u8>): vector<u8> {
         let public_key = vector::empty<u8>();
-        let i = scheme_length() + signature_length();
-        let public_key_position = scheme_length() + signature_length() + public_key_length();
+        let i = auth_validator_id_length() + signature_length();
```
[the remark this answers is not in the document]
Comment (feliciss) [reply crates/rooch-framework/sources/crypto/ed25519.move]: Added a TODO.

File: crates/rooch-framework/sources/crypto/ed25519.move [hunk 16]
```diff
@@ -21,8 +21,8 @@ module rooch_framework::ed25519 {
 
     public fun get_public_key_from_authenticator_payload(authenticator_payload: &vector<u8>): vector<u8> {
```
[the remark this answers is not in the document]
Comment (feliciss) [reply crates/rooch-framework/sources/crypto/ed25519.move]: Moved.

File: crates/rooch-framework/sources/crypto/ecdsa_k1.move [hunk 17]
```diff
@@ -44,8 +44,8 @@ module rooch_framework::ecdsa_k1 {
 
     public fun get_public_key_from_authenticator_payload(authenticator_payload: &vector<u8>): vector<u8> {
```
[the remark this answers is not in the document]
Comment (feliciss) [reply crates/rooch-framework/sources/crypto/ecdsa_k1.move]: Removed.

File: crates/rooch-framework/sources/address_mapping.move [hunk 18]
```diff
@@ -10,15 +10,15 @@ module rooch_framework::address_mapping{
...
+    //The multichain id standard is defined in [slip-0044](https://github.com/satoshilabs/slips/blob/master/slip-0044.md)
     //Please keep consistent with rust Symbol
-    const COIN_ID_BITCOIN: u64 = 0;
-    const COIN_ID_ETHER: u64 = 60;
-    const COIN_ID_NOSTR: u64 = 1237;
-    const COIN_ID_ROOCH: u64 = 20230101;
+    const MULTICHAIN_ID_BITCOIN: u64 = 0;
+    const MULTICHAIN_ID_ETHER: u64 = 60;
+    const MULTICHAIN_ID_NOSTR: u64 = 1237;
+    const MULTICHAIN_ID_ROOCH: u64 = 20230103;
```
[the remark this answers is not in the document]
Comment (feliciss) [reply crates/rooch-framework/sources/address_mapping.move]: Changed.

**Review by jolestar — APPROVED** [review]

**Pull request merged**